Updated CE verification script to handle job router transforms (HTCONDOR-527)

markcoatsworth opened the pull request.

brianhlin commented on src/verify_ce_config.py:
```diff
@@ -103,35 +108,45 @@ def main():
-                 + "may not have any effect. Use the 'set_' prefix instead.")
+    # Also warn about routes specified in JOB_ROUTER_ROUTE_<name> transforms that don't appear in the parsed JRE.
+    if jr_transforms:
+        for transform_attr, transform_value in jr_transforms:
+            missing_route_def = set(transform_value.replace(',',' ').split()).difference(set(parse_route_names(config_val)))
+            if missing_route_def:
+                warn("The following are specified in %s "
+                        "but cannot be found in JOB_ROUTER_ENTRIES: %s"
+                        % transform_attr, ', '.join(missing_route_def))
```
This isn't quite the right check to make (transforms are separate from entries). We should warn if there are route names showing up in `JOB_ROUTER_ROUTE_NAMES` that don't have a corresponding `JOB_ROUTER_ROUTE_<NAME>` config variable

Also, the usage of `config_val` will likely lead to confusion down the line since it's defined in the for loop and it luckily works here since the last iteration in the loop will set `config_val` the ads defined by `JOB_ROUTER_ENTRIES`

brianhlin commented on src/verify_ce_config.py:
```diff
@@ -63,27 +66,29 @@ def main():
-            # non-string/non-file output and htcondor.param shouldn't contain such values
-            debug("Failed to parse %s configuration value. " % attr +
-                  "Can be safely ignored if using 'JOB_ROUTER_ROUTE_*' ClassAd transforms.")
-
-        # If JRD or JRE can't be parsed, the job router can't function
-        if not parsed_jr_ads[attr]:
-            debug("Could not read %s in the HTCondor-CE configuration. " % attr +
-                  "Can be safely ignored if using 'JOB_ROUTER_ROUTE_*' ClassAd transforms.")
+        # We can ignore this for JOB_ROUTER_ENTRIES if there's at least one JOB_ROUTER_ROUTE_<name>
+        if attr != "JOB_ROUTER_ENTRIES" or not jr_transforms:
+            try:
+                parsed_jr_ads[attr] = list(classad.parseAds(config_val))
+            except ValueError:
+                # We shouldn't ever get here since classad.parseAds() only raises ValueError when it's given
+                # non-string/non-file output and htcondor.param shouldn't contain such values
+                debug("Failed to parse %s configuration value. " % attr)
+
+                # If JRD or JRE can't be parsed, the job router can't function
+                if not parsed_jr_ads[attr]:
+                    debug("Could not read %s in the HTCondor-CE configuration. " % attr)
```
I think we should revert this bit

brianhlin reviewed (changes requested):
Hey @markcoatsworth , a few comments but overall, I think we should revisit that refactoring idea that you had and just simplify it. For everyone's sanity, it makes sense to treat the new transform config and old classad config as mutually exclusive, i.e. if `JOB_ROUTER_ROUTE_*` is defined, only do JRT validations.

So I propose that we:

1.  Only run all of the old checks `if not jr_transform`
2.  If they have any JRTs defined, warn about names in `JOB_ROUTER_ROUTE_NAMES` that don't have a corresponding JRT
3. Since `main()` already has too many branches, split all of the old checks into a separate function

markcoatsworth commented:
My most recent commit (https://github.com/htcondor/htcondor-ce/pull/497/commits/4d9241d0e933692eba573f0bedd1a71c012752ce) addresses all these comments but it's also not showing up in this window? Let me know if there's anything I need to do on my end.

brianhlin commented on src/verify_ce_config.py:
```diff
@@ -63,75 +66,85 @@ def main():
+    # If JOB_ROUTER_ROUTE_<name> rules are defined, verify that these transforms exist
+    if jr_transforms:
+        route_names = htcondor.param.get("JOB_ROUTER_ROUTE_NAMES", "")
+        for name in route_names.split(" "):
```
```suggestion
        for name in re.split(r'[, \t]+', route_names):
```

`JOB_ROUTER_ROUTE_NAMES` can be comma separated

brianhlin commented on src/verify_ce_config.py:
```diff
@@ -63,75 +66,85 @@ def main():
+    # If JOB_ROUTER_ROUTE_<name> rules are defined, verify that these transforms exist
+    if jr_transforms:
+        route_names = htcondor.param.get("JOB_ROUTER_ROUTE_NAMES", "")
+        for name in route_names.split(" "):
+            route_def = htcondor.param.get(f"JOB_ROUTER_ROUTE_{name}", "")
+            if not route_def:
+                warn(f"The route {name} is specified in JOB_ROUTER_ROUTE_NAMES, but no corresponding JOB_ROUTER_ROUTE_{name} exists")
```
You could probably do this with a `set.difference()` but the list of routes are probably not large enough to matter

markcoatsworth replied:
Let's just leave as-is for now?

brianhlin replied:
Yeah, I think it's fine

brianhlin reviewed (approved):
LGTM except some minor comments

brianhlin commented on src/verify_ce_config.py:
```diff
@@ -63,27 +66,29 @@ def main():
 
     # Create dict whose values are lists of ads specified in the relevant JOB_ROUTER_* variables
     parsed_jr_ads = defaultdict(list)
+    jr_transforms = {key : value for key, value in htcondor.param.items() if "JOB_ROUTER_ROUTE_" in key}
```
Just tested this, it looks like the keys in `htcondor.param` are case insensitive

markcoatsworth commented:
I think this is good to go, over to you to merge.
